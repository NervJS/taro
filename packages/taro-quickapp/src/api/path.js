// copy from https://github.com/jinder/path/blob/master/path.js
export function resolve () {
  var resolvedPath = ''
  var resolvedAbsolute = false

  for (var i = arguments.length - 1; i >= -1 && !resolvedAbsolute; i--) {
    var path = (i >= 0) ? arguments[i] : '/'

    if (!path) {
      continue
    }

    resolvedPath = path + '/' + resolvedPath
    resolvedAbsolute = path[0] === '/'
  }

  // At this point the path should be resolved to a full absolute path, but
  // handle relative paths to be safe (might happen when process.cwd() fails)

  // Normalize the path
  resolvedPath = normalizeArray(resolvedPath.split('/'),
    !resolvedAbsolute).join('/')

  return ((resolvedAbsolute ? '/' : '') + resolvedPath) || '.'
}

function normalizeArray (parts, allowAboveRoot) {
  var res = []
  for (var i = 0; i < parts.length; i++) {
    var p = parts[i]

    // ignore empty parts
    if (!p || p === '.') { continue }

    if (p === '..') {
      if (res.length && res[res.length - 1] !== '..') {
        res.pop()
      } else if (allowAboveRoot) {
        res.push('..')
      }
    } else {
      res.push(p)
    }
  }

  return res
}
