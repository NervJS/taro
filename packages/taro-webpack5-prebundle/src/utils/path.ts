export const addLeadingSlash = (url = '') => (url.charAt(0) === '/' ? url : '/' + url)
export const addTrailingSlash = (url = '') => (url.charAt(url.length - 1) === '/' ? url : url + '/')
export function parsePublicPath (publicPath = '/') {
  return ['', 'auto'].includes(publicPath) ? publicPath :  addTrailingSlash(publicPath)
}