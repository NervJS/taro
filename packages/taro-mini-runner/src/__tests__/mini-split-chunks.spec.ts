import * as fs from 'fs-extra'
import * as path from 'path'

let compile
let readDir

describe('mini-split-chunks', () => {
  beforeAll(() => {
    jest.unmock('webpack')
  })

  beforeEach(() => {
    jest.isolateModules(() => {
      const compiler = require('./utils/compiler')
      compile = compiler.compile
      readDir = compiler.readDir
    })
  })

  test('should process mini-split-chunks', async () => {
    const appName = 'mini-split-chunks'
    const { stats, config } = await compile(appName, {
      optimizeMainPackage: {
        enable: true,
        exclude: [
          path.resolve(__dirname, './fixtures/mini-split-chunks/src/utils/testExcludeString.js'),
          (module) => module.resource?.indexOf('testExcludeFunction') >= 0
        ]
      },
      postcss: {
        cssModules: {
          enable: true,
          config: {
            namingPattern: 'module',
            generateScopedName: '[name]__[local]___[hash:base64:5]'
          }
        }
      }
    })

    const assets = stats.toJson().assets || []
    expect(assets.length).toMatchSnapshot()

    const files = readDir(fs, config.outputRoot)

    /**
     * 因为该plugin输出的文件名跟用户路径有关，snapshot输出的文件内容不稳定，这里改用判断输出文件是否存在来测试
     */
    files.forEach(file => {
      const fileContent = fs.readFileSync(file).toString()
      const jsExt = '.js'
      const wxssExt = '.wxss'
      const needTestPages = [
        'packageA/detail/index',
        'packageA/my/index',
        'packageB/list/index'
      ]
      const checkChunksExist = ({ ext, regexp }) => {
        const isNeedTestFile = needTestPages.find(needTestPage => new RegExp(`${needTestPage}${ext}$`).test(file))

        if (!isNeedTestFile) return

        const matchChunks = (fileContent || '').match(new RegExp(regexp, 'g'))
        const chunkPaths = (matchChunks || []).map(chunkPath => {
          return chunkPath.replace(new RegExp(regexp), '$1')
        })

        chunkPaths.forEach(chunkPath => {
          const chunkExt = ext === jsExt ? jsExt : ''
          const chunkAbsolutePath = path.resolve(file, '..', chunkPath + chunkExt)
          const isExists = fs.pathExistsSync(chunkAbsolutePath)

          // 根据match到的module path查看输出的chunk是否存在
          expect(isExists).toBe(true)
        })
      }

      // js
      if (path.extname(file) === jsExt) {
        checkChunksExist({
          ext: jsExt,
          regexp: 'require\\("(.*)"\\);'
        })
      }

      // wxss
      if (path.extname(file) === wxssExt) {
        checkChunksExist({
          ext: wxssExt,
          regexp: new RegExp('@import "(.*)";')
        })
      }
    })
  })
})
