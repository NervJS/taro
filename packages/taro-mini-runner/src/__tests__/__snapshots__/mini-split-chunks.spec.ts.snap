// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mini-split-chunks should process mini-split-chunks 1`] = `40`;
